[Agent] Allow CA cert pinning on the Elasticsearch output or any code that user tlscommon.TLSConfig builder.

ph opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
-->

## What does this PR do?

When multiples CA are presents on the system we cannot ensure that a
specific one was used to validate the chain exposed by the server.
This PRs adds a `ca_sha256` option to the `tlscommon.TLSConfig` that is used by all
the code that has to create a TCP client with TLS support.

When the option is set, it will hook a new callback in the validation
chains that will inspect the verified and validated chains by Go to
ensure that a certificate in the chains matches the provided pin.

A usage example for the Elasticsearch output.

```yaml
output.elasticsearch:
  hosts: [127.0.0.1:9200]
  ssl.ca_sha256: <base64_encoded_sha256>
```

You can generate the pin using the **openssl** binary with the
following command:

```
openssl x509 -in ca.crt -pubkey -noout | openssl pkey -pubin -outform der | openssl dgst -sha256 -binary | openssl enc -base64
```

OpenSSL's [documentation](https://www.openssl.org/docs/manmaster/man1/dgst.html)

You will need to start Elasticsearch with the following options

```yaml
xpack.security.enabled: true
indices.id_field_data.enabled: true
xpack.license.self_generated.type: trial
xpack.security.http.ssl.enabled: true
xpack.security.http.ssl.key: /etc/pki/localhost/localhost.key"
xpack.security.http.ssl.certificate: /etc/pki/localhost/localhost.crt"
xpack.security.http.ssl.certificate_authorities: /etc/pki/ca/ca.crt"
```

This pull request also includes a new service in the docker-compose.yml
that will start a new Elasticsearch server with TLS and security
configured.

<!-- Mandatory
Explain here the changes you made on the PR. Please explain the WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
-->


## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [X] I have added tests that prove my fix is effective or that my feature works

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [x] Need to modify the configuration reference.
- [x] Need to add the documentation.

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

There are multiple ways to tests this, you can use the _elasticsearch_ssl_ service defined in the docker-compose file that will start a configured Elasticsearch with TLS and the right CA/Certificate configured. 

Or you can generate a _CA_ and a _server certificate_ using the [elasticsearch-certutil](https://www.elastic.co/guide/en/elasticsearch/reference/current/certutil.html), make sure to export them in the PEM format. 


In both case you can use the following _openssl_ command to generate the pin from the CA.

```
openssl x509 -in ca.crt -pubkey -noout | openssl pkey -pubin -outform der | openssl dgst -sha256 -binary | openssl enc -base64
````

And use the following output configuration:

```
output.elasticsearch:
  hosts: [127.0.0.1:9200]
  username: <username>
  password: <password>
  ssl.ca_sha256: <base64_encoded_sha1>
  ssl.certificate_authorities: <path_to_ca>
```


## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123

- Relates #123

- Requires #123

- Superseds elastic/beats#123
-->

- Fixes:  elastic/beats#15717
- Relates: #15718, elastic/beats#15716

michalpristas commented on testing/environments/docker/elasticsearch/pki/generate_pki.sh:
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
+go run ../../../../../dev-tools/cmd/pin_cacert/main.go -cacert ca/ca.crt
```
i was looking if we have some ENV which can be useful here instead of ugly `../` like `{ES_BEATS}`maybe?

ph replied:
The problem is not available unless you set it up. This is more like an helper methods.

ph replied:
I fixed it another way.. I used openssl binaries to generate the pin so the path is now gone.

michalpristas commented on libbeat/common/transport/tlscommon/ca_pinning.go:
```diff
@@ -0,0 +1,71 @@
+		// The chain of trust has been already established before the call to the VerifyPeerCertificate
+		// function, after we go through the chain to make sure we have CA certificate that match the provided
+		// pin.
+		for _, chain := range verifieds {
```
small nit: name `verifieds` looks weird

ph replied:
going to replace it with `verifiedChains`

elasticmachine commented:
Pinging @elastic/ingest (Project:fleet)

ph commented:
@michalpristas I have addresses your concerns and made the PR reader for review.

ph commented on libbeat/common/transport/tlscommon/ca_pinning.go:
```diff
@@ -0,0 +1,71 @@
+
+		return ErrCAPingMatch
+	}
+}
```
As discussed from last week @scunningham I ping you on the pull request.

scunningham commented on libbeat/common/transport/tlscommon/ca_pinning.go:
```diff
@@ -0,0 +1,71 @@
+
+type pins []string
+
+func (p *pins) Matches(candidate string) bool {
```
Does this need to be a pointer receiver?

scunningham commented on libbeat/common/transport/tlscommon/ca_pinning.go:
```diff
@@ -0,0 +1,71 @@
+		// The chain of trust has been already established before the call to the VerifyPeerCertificate
+		// function, after we go through the chain to make sure we have CA certificate that match the provided
+		// pin.
+		for _, chain := range verifiedChains {
```
So I read this as matching any cert in any verified chains.  Not necessarily the root CA.  Correct?

ph replied:
Yes, but I could restrict it on the first item of the verified chain, would that be better for you?

scunningham commented on libbeat/common/transport/tlscommon/ca_pinning.go:
```diff
@@ -0,0 +1,71 @@
+// Fingerprint takes a certificate and create a hash of the DER encoded public key.
+func Fingerprint(certificate *x509.Certificate) string {
+	// uses the DER encoded version of the public key to generate the pin.
+	hash := sha256.Sum256(certificate.RawSubjectPublicKeyInfo)
```
was expecting fingerprint as typically calculated in openssl, but the way you did it is smaller

openssl x509 -sha256 -fingerprint -in ca.crt -noout
SHA256 Fingerprint=4F:3C:B4:4E:BB:2B:3C:02:4F:DA:95:EA:81:19:3F:F5:76:75:D4:D6:C6:C1:A3:B9:46:57:24:AC:37:56:39:70

ph replied:
I was using the [hpkp](https://developer.mozilla.org/en-US/docs/Web/HTTP/Public_Key_Pinning) strategy to make it smaller, I don't have a strong argument for going that route other than might be a bit better on the CLI. What do you think?

scunningham replied:
Smaller is better; I like the way you did it..

ph commented on x-pack/filebeat/filebeat.yml:
```diff
@@ -21,7 +21,7 @@ filebeat.inputs:
 - type: log
 
   # Change to true to enable this input configuration.
-  enabled: false
```
I need to remove this :man_facepalming:

ph commented:
@urso When this goes it this will affect any client that uses `tlscommon.TLSConfig` I think its a good idea that you take a look.

ph commented:
I have added the test-plan labels, but we have integration tests in this PR so this might not be necessary.

urso commented on libbeat/common/transport/tlscommon/ca_pinning.go:
```diff
@@ -0,0 +1,69 @@
+)
+
+// ErrCAPingMatch is returned when no pin is matched in the verified chain.
+var ErrCAPingMatch = errors.New("provided CA certificate pins doesn't match any of the certificate authorities used to validate the certificate")
```
s/ErrCAPingMatch/ErrCAPinMissmatch/ ?

urso commented on libbeat/common/transport/tlscommon/tls_config.go:
```diff
@@ -79,17 +83,25 @@ func (c *TLSConfig) ToConfig() *tls.Config {
+		CurvePreferences:      c.CurvePreferences,
+		Renegotiation:         c.Renegotiation,
+		ClientAuth:            c.ClientAuth,
+		VerifyPeerCertificate: verifyPeerCertFn,
```
the callback (the way we use it) requires `insecure` to be false. Otherwise the callback will get an empty chain. This is fine, but we should document somewhere (code and or docs), that pinning adds an additional check to normal verification, but is no replacement for e.g. host checks.

ph replied:
I have added a comment in both the go code and the user documentation.

ph commented:
jenkins test this please


Two host with error connecting :(

ph commented:
Rebased and added a bit more time to get the ES container healthy, I've took the logic from the `x-pack/libbeat` healthcheck, which was more resilient.

ph commented:
Jenkins test this please

ruflin commented on CHANGELOG.next.asciidoc:
```diff
@@ -52,6 +52,8 @@ https://github.com/elastic/beats/compare/v7.0.0-alpha2...master[Check the HEAD d
 - Fix issue where default go logger is not discarded when either * or stdout is selected. {issue}10251[10251] {pull}15708[15708]
 - Upgrade go-ucfg to latest v0.8.1. {pull}15937{15937}
 - Fix index names for indexing not always guaranteed to be lower case. {pull}16081[16081]
+- Remove superfluous use of number_of_routing_shards setting from the default template. {pull}16038[16038]
```
Merge issue?

ph commented:
updated changelog, merging.
